src/vulkan/swap_chain.cpp

```diff
@@ -13,20 +13,13 @@
 #include <stdexcept>
 
 namespace IC {
-    SwapChain::SwapChain(VulkanDevice &deviceRef, VkExtent2D extent) : _device{deviceRef}, _windowExtent{extent} {
-        Init();
-    }
-
     SwapChain::SwapChain(VulkanDevice &deviceRef, VkExtent2D extent, std::shared_ptr<SwapChain> previous)
-        : _device{deviceRef}, _windowExtent{extent}, _oldSwapChain{previous} {
-        Init();
-
-        // clean up old swap chain as it's no longer needed
-        _oldSwapChain = nullptr;
+        : _device{deviceRef}, _windowExtent{extent} {
+        Init(previous);
     }
 
-    void SwapChain::Init() {
-        CreateSwapChain();
+    void SwapChain::Init(std::shared_ptr<SwapChain> previous) {
```

Comment:
This is fine as is, but want to point out passing a std::shared_ptr by value to each function leads to additional memory writes. This is due to having to construct a copy of the object for every function, leading to additions/subtractions of the shared pointer's reference count. Almost always is a std::shared_ptr passed by reference.